main.js

```diff
@@ -1,118 +1,90 @@
+
+"use strict";
 
-const {devtools} = Cu.import("resource://gre/modules/devtools/Loader.jsm", {});
+const { Cu } = require("chrome");
+const { devtools } =
+  Cu.import("resource://gre/modules/devtools/Loader.jsm", {});
 const devtoolsRequire = devtools.require;
-const {ConnectionManager} = devtoolsRequire("devtools/client/connection-manager");
-let {Devices} = Cu.import("resource://gre/modules/devtools/Devices.jsm");
-const {when: unload} = require("sdk/system/unload");
+const { defineLazyGetter } = devtoolsRequire("devtools/toolkit/DevToolsUtils");
 
-let promise;
-try {
-  promise = Cu.import("resource://gre/modules/commonjs/promise/core.js").Promise;
-} catch (e) {
-  promise = Cu.import("resource://gre/modules/commonjs/sdk/core/promise.js").Promise;
-}
+const events = require("sdk/event/core");
+const { when: unload } = require("sdk/system/unload");
+const system = require("sdk/system");
+const { Devices } = Cu.import("resource://gre/modules/devtools/Devices.jsm");
+const { gDevToolsBrowser } =
+  Cu.import("resource:///modules/devtools/gDevTools.jsm");
+defineLazyGetter(this, "adb", () => {
+  return require("./adb");
+});
+defineLazyGetter(this, "fastboot", () => {
+  return require("./fastboot");
+});
+defineLazyGetter(this, "Device", () => {
+  return require("./device");
+});
+
+// Set this right away on startup
+Devices.helperAddonInstalled = true;
+unload(() => Devices.helperAddonInstalled = false);
 
-// As of Firefox 36, WebIDE exposes an API to register new runtimes.
-try {
+function onADBStart() {
+  // As of Firefox 36, WebIDE exposes an API to register new runtimes.
   let Runtimes = devtoolsRequire("devtools/webide/runtimes");
   if (Runtimes && Runtimes.RuntimeScanners) {
     let scanner = require("./scanner");
     scanner.register();
   }
-} catch (e) {}
-
-Devices.helperAddonInstalled = true;
-exports.shutdown = function() {
-  Devices.helperAddonInstalled = false;
-  adb.stop(true);
-};
 
-adb.start().then(function () {
-  adb.trackDevices();
-});
+  adb.start().then(function() {
+    adb.trackDevices();
+  });
+}
+Devices.on("adb-start-polling", onADBStart);
+unload(() => Devices.off("adb-start-polling", onADBStart));
 
 /**
- * A Device instance is created and registered with the Devices module whenever
- * ADB notices a new device is connected.
- *
- * Any changes here should be examined carefully for backwards compatibility
- * issues.  Other add-ons, like the Tools Adapter, make use of these low-level
- * device objects too.
+ * For Firefox 41 and earlier, WebIDE does not know to emit "adb-start-polling"
+ * to enable this add-on lazily.  So, we check for WebIDE startup using a
+ * promise added in Firefox 39.
  */
-function Device(id) {
-  this.id = id;
+let version = Number.parseInt(system.version.split(".")[0]);
+if (system.id == "{ec8030f7-c20a-464f-9b0e-13a3a9e97384}" && version < 42 &&
+    gDevToolsBrowser.isWebIDEInitialized) {
+  gDevToolsBrowser.isWebIDEInitialized.promise.then(onADBStart);
```

Comment:
:(
I thought that `isWebIDEInitialized` was introduced later that 39...
We would still miss current 38 release.
What about:
```
if (system.id == "{ec8030f7-c20a-464f-9b0e-13a3a9e97384}" && version < 42)
  if (gDevToolsBrowser.isWebIDEInitialized) {
    gDevToolsBrowser.isWebIDEInitialized.promise.then(onADBStart);
  } else {
   onADBStart();
  }
}
```